File: backend/src/types/sync-errors.ts
export type SyncStage = "enumeration" | "fetch" | "normalize" | "index" | "checkpoint";

export const SyncErrorCode = {
  ConnectorNotFound: "CONNECTOR_NOT_FOUND",
  InvalidConfig: "INVALID_CONFIG",
  Unauthorized: "UNAUTHORIZED",
  EnumerationFailed: "ENUMERATION_FAILED",
  FetchFailed: "FETCH_FAILED",
  EmptyContent: "EMPTY_CONTENT",
  ParseFailed: "PARSE_FAILED",
  IndexingFailed: "INDEXING_FAILED",
  CheckpointFailed: "CHECKPOINT_FAILED",
  Unknown: "UNKNOWN",
} as const;

export type SyncErrorCodeValue = (typeof SyncErrorCode)[keyof typeof SyncErrorCode];

export class SyncPipelineError extends Error {
  code: SyncErrorCodeValue;
  stage: SyncStage;
  retriable: boolean;

  constructor(options: {
    code: SyncErrorCodeValue;
    message: string;
    stage: SyncStage;
    retriable?: boolean;
    cause?: unknown;
  }) {
    super(options.message);
    this.name = "SyncPipelineError";
    this.code = options.code;
    this.stage = options.stage;
    this.retriable = options.retriable ?? false;
    if (options.cause !== undefined) {
      (this as Error & { cause?: unknown }).cause = options.cause;
    }
  }
}

export function toSyncPipelineError(error: unknown, fallbackStage: SyncStage): SyncPipelineError {
  if (error instanceof SyncPipelineError) return error;
  const message = error instanceof Error ? error.message : "Unknown sync error";
  return new SyncPipelineError({
    code: SyncErrorCode.Unknown,
    stage: fallbackStage,
    message,
    retriable: false,
    cause: error,
  });
}

File: apps/api/app/types/sync_errors.py
from __future__ import annotations

from typing import Literal


SyncStage = Literal["enumeration", "fetch", "normalize", "index", "checkpoint"]


class SyncErrorCode:
    CONNECTOR_NOT_FOUND = "CONNECTOR_NOT_FOUND"
    INVALID_CONFIG = "INVALID_CONFIG"
    UNAUTHORIZED = "UNAUTHORIZED"
    ENUMERATION_FAILED = "ENUMERATION_FAILED"
    FETCH_FAILED = "FETCH_FAILED"
    EMPTY_CONTENT = "EMPTY_CONTENT"
    PARSE_FAILED = "PARSE_FAILED"
    INDEXING_FAILED = "INDEXING_FAILED"
    CHECKPOINT_FAILED = "CHECKPOINT_FAILED"
    UNKNOWN = "UNKNOWN"


class SyncPipelineError(Exception):
    def __init__(
        self,
        *,
        code: str,
        message: str,
        stage: SyncStage,
        retriable: bool = False,
        cause: BaseException | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.stage = stage
        self.retriable = retriable
        if cause is not None:
            self.__cause__ = cause


def to_sync_pipeline_error(error: BaseException, fallback_stage: SyncStage) -> SyncPipelineError:
    if isinstance(error, SyncPipelineError):
        return error
    return SyncPipelineError(
        code=SyncErrorCode.UNKNOWN,
        stage=fallback_stage,
        message=str(error),
        retriable=False,
        cause=error,
    )
